Introduce Paste Tools and Paste from GDocs

Comandeer opened the pull request:
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What is the proposed changelog entry for this pull request?

```
* [#835](https://github.com/ckeditor/ckeditor-dev/issues/835): Added new `pastetools` and `pastefromgdocs` plugins.
```

## What changes did you make?

I've introduced `pastetools` plugin, which allows registering paste handlers:

```javascript
class PasteHandler {
	// Array of external filters that have to be loaded before executing current paste handler handle method.
	filters: string[]

	// Checks if handler can handle current paste event.
	canHandle( evt ): boolean

	// Actual paste handling
	handle( evt ): void
}
```

I've refactored `pastefromword` to use the new mechanism. I've also split PfW filter into two:

* `pastefromword/filter/default.js` – Word specific stuff,
* `pastetools/filter/common.js` – stuff that can be useful also in other filters.

Additionally I've introduced simple API for creating filters: `CKEDITOR.plugins.pastetools.createFilter`. Every filter definition is divided into two parts:

* `rules` – array of functions that will return [filter rules](http://ckeditor.test/ckeditor-docs/build/docs/ckeditor4/latest/api/CKEDITOR_htmlParser_filterRulesDefinition.html). Every rule-creating function receives `html` (pasted HTML), `editor` and `filter` parameters.
* `additionalTransforms` – function that transforms pasted HTML before passing it into filter. It receives `html` and `editor` parameters.

I've also added `pastefromgdocs` plugin, with its own filter and paste handler.

Currently `pastefromgdocs` does not have any tests, except the generic manual one. 

I've also had to modify some PfW tests:

* `generated/functions.js` – part of it moved to `pastetools` directory,
* `generated/heuristics.js`,
* `helpers.js`
* `image.js`
* `parsestyles.js` – moved to `pastetools` directory
* `pasteimage.js`.

Some API docs are missing or not updated.

There is also small issue with config variables. Currently some parts of common filter (`CKEDITOR.plugins.pastetools.filters.common.styles`) use some Paste from Word config variables:

* `config.pasteFromWordRemoveFontStyles`,
* `config.pasteFromWord_keepZeroMargins`.

I'm thinking about introducing their pastetools counterparts:

* `config.pasteTools_removeFontStyles`,
* `config.pasteTools_keepZeroMargins`.

It would also mean that PfW variables become deprecated and if user set both of them, `pastetools` ones will take precedence.

I'm also thinking about namespaces. Currently `CKEDITOR.plugins.pastetools.filters` is used to store filters' config and `CKEDITOR.pasteFilters` is used to store filters' instances. However it introduces superlong namespaces (`CKEDITOR.plugins.pastetools.filters.common.styles.normalizedStyles` – ouch…). Not sure if it's fixable in some really sensible way?

I'm also not a big fan of `pastetools` name. Maybe something like `pastehandler`, `specialpaste`, maybe even `uberpaste`? `pastetools` isn't really about tools for pasting…

Big part of this PR is based on #2472 by @jacekbogdanski, so big kudos to him, as I could skip a huge part of initial research! 👏👏👏Too bad that simple cherry-picking wasn't possible due to the fact how much current approach is different from the one in mentioned PR.

Closes #835.

Comandeer commented:
There is also one more important change from the previous PR: Paste from GDocs is no longer a part of Paste from Word, therefore it doesn't fire `pastefromword` or `afterpastefromword` events. Additionally it doesn't have its own paste button. Filtering GDocs content is totally transparent for the user.

msamsel commented on plugins/pastefromgdocs/plugin.js:
```diff
@@ -0,0 +1,47 @@
+				],
+
+				canHandle: function( evt ) {
+					return evt.data.dataValue.match( /id=(\"|\')docs\-internal\-guid\-/ );
```
Why there is used `String.match` not `Regexp.test`? This statement seems to determine a boolean state and a result of a match seems to not be important.
SO claims that using `test` is faster for such cases:
https://stackoverflow.com/questions/10940137/regex-test-v-s-string-match-to-know-if-a-string-matches-a-regular-expression

Comandeer replied:
TBH without reason – this logic was copied from the previous PR and I didn't see a need of changing it (as it was working). But yeah, I can refactor it later.

Comandeer commented:
I've introduced middleware architecture described in #3258.

Comandeer commented:
Introducing fixture-based tests revealed that filter has one visible flaw: it removes elements with `[id^=docs-internal-guid-]`. However in Firefox and Edge such element also contains styles and without them pasted content has changed layout. I'll provide fix in the nearest days.

Comandeer commented:
I've updated rules for dealing with elements with GDocs specific id. Instead of removing these elements, I just remove their ids.

I've also added rule to remove all comments from the code.

f1ames commented:
Also extracted follow-ups:

* Documentation for PFGD - #3274 
* Documentation for `pastetools`- #3275
* Unify pasting UI (mentioned in above review in UI section) - #3276

msamsel commented:
During work on Past from Google Docs in CKE5, I noticed a few cases which maybe you would also like to check in CKE4.
1. Content in **Safari** like this one, after <kbd>cmd</kbd> + <kbd>a</kbd>,  <kbd>cmd</kbd>  + <kbd>c</kbd> and  <kbd>cmd</kbd>  + <kbd>v</kbd> in editor doesn't contain `guid-...` so might won't be recognized as content from google docs. It might be importnatn if there is no general content fixer regardless of data source.
![image](https://user-images.githubusercontent.com/9017322/62301043-1ef5a100-b478-11e9-9446-e59c550aef39.png)
2. There are cases with lists which have invalid html syntax. Things worth checking:
  * https://github.com/ckeditor/ckeditor5-paste-from-office/pull/72#issuecomment-516458324
  * partially selected list for example like this one:
![image](https://user-images.githubusercontent.com/9017322/62301466-fc17bc80-b478-11e9-8a38-fa4dec2e1d36.png)

Comandeer commented on plugins/pastetools/plugin.js:
```diff
@@ -0,0 +1,190 @@
+	// Join all filters in one big array and then filter out duplicates.
+	function getFilters( handlers ) {
+		var filters = CKEDITOR.tools.array.reduce( handlers, function( filters, handler ) {
+			if ( !CKEDITOR.tools.array.isArray( handler.filters ) ) {
```
Yes.

Comandeer commented on plugins/pastefromgdocs/plugin.js:
```diff
@@ -0,0 +1,51 @@
+				filters: [
+					CKEDITOR.getUrl( pasteToolsPath + 'filter/common.js' ),
+					CKEDITOR.getUrl(  path + 'filter/default.js' )
+				],
```
IMO passing URLs is far more flexible and allows user to store filters in some non-standard locations. Such locations can be present in CMSes.

Comandeer commented on tests/plugins/pastefromword/generated/_helpers/pfwTools.js:
```diff
@@ -124,6 +124,21 @@
 			return tests;
+		},
+
+		loadFilters: function loadFilters( filters, callback ) {
```
No, I did it this way to avoid complicated bender imports. However it can be moved to `ptTools` and removed from `pfwTools`.

Comandeer commented:
> I think that would be a good direction. However, I'm not sure if we should deprecated the Word ones as one may still want to use them only for pasting from Word and not from other office-like applications. Do you think this might be common use case? Or we can deprecate, but mention that they still can be used if one wants to affect only Word input.

I'm afraid that it's not possible with our current architecture, as all filters connected with fonts are moved to the shared filter. It will have to know that the content being filtered is from Word, but I don't know if there is any easy way to pass there such info 🤔

Comandeer commented on plugins/pastefromgdocs/filter/default.js:
```diff
@@ -0,0 +1,104 @@
+		};
+	};
+
+	function falseIfEmpty( value ) {
```
As it could be reused as indicated in https://github.com/ckeditor/ckeditor-dev/pull/3257#discussion_r303841684, I won't inline it.

Comandeer commented on plugins/pastefromgdocs/filter/default.js:
```diff
@@ -0,0 +1,104 @@
+	 * @private
+	 * @member CKEDITOR.plugins.pastetools.filters.gdocs
+	 */
+	plug.rules = function( html, editor, filter ) {
```
Such method was used in the original PR. I've used it here without much reflection TBH. I'll simplify it.

Comandeer commented on plugins/pastefromgdocs/filter/default.js:
```diff
@@ -0,0 +1,104 @@
+		return value;
+	}
+
+	CKEDITOR.pasteFilters.gdocs = pastetools.createFilter( {
```
I wanted to separate actual filters from helper functions and settings. However if you think that it can be simplified, we can think of something suitable. Maybe `CKEDITOR.plugins.pastetools.filters.gdocs.filter`/`CKEDITOR.plugins.pastetools.filters.gdocs.instance` or `CKEDITOR.plugins.pastetools.filters.instances.gdocs`? But it's probably clear now why I chose something different 😄

Comandeer commented on plugins/pastefromgdocs/plugin.js:
```diff
@@ -0,0 +1,51 @@
+					if ( editor.config.forcePasteAsPlainText === true ) {
+						// If `config.forcePasteAsPlainText` set to true, force plain text even on Word content (#1013).
+						data.type = 'text';
+					} else if ( !CKEDITOR.plugins.clipboard.isCustomCopyCutSupported && editor.config.forcePasteAsPlainText === 'allow-word' ) {
```
I'm not sure TBH. Maybe we can introduce some other values? E.g. `allow-gdocs` or `allow-pastetools` (to allow formatting from everything handled by `pastetools`).

Comandeer replied:
https://github.com/ckeditor/ckeditor-dev/issues/3369

Comandeer commented on plugins/pastefromgdocs/plugin.js:
```diff
@@ -0,0 +1,51 @@
+						// In browsers using pastebin when pasting from Word, evt.data.type is 'auto' (not 'html') so it gets converted
+						// by 'pastetext' plugin to 'text'. We need to restore 'html' type (#1013) and (#1638).
+						data.type = 'html';
+					}
```
Yes, it should be a generic handler, but it's a little bit tricky.

At the moment it's not possible to detect the source of the paste, as this information is not passed in the event. Ideally we can introduce the unified handling for forcing paste type and passing info about the source of paste, e.g.:

```javascript
// PfW handler
editor.pasteTools.register( {
	priority: 10,
	handle( { data } ) {
		if ( isWordContent || editor._.nextPasteType ) {
			data.source = 'word';
		}
	}
} );

// Generic handler in pastetext.
editor.pasteTools.register( {
	priority: 999, // Fire at the end.
	handle( { data } ) {
		if ( editor.config.forcePasteAsPlainText === true ) {
			// If `config.forcePasteAsPlainText` set to true, force plain text even on Word content (#1013).
			data.type = 'text';
		} else if ( !CKEDITOR.plugins.clipboard.isCustomCopyCutSupported && editor.config.forcePasteAsPlainText === 'allow-word' && data.source === 'word' ) {
			// In browsers using pastebin when pasting from Word, evt.data.type is 'auto' (not 'html') so it gets converted
			// by 'pastetext' plugin to 'text'. We need to restore 'html' type (#1013) and (#1638).
			data.type = 'html';
		}
	}
} );

// Generic handler in pastetools.
editor.pasteTools.register( {
	priority: Infinity, // Fire at THE VERY END.
	handle( evt ) {
		delete editor._.nextPasteType;
	}
} );
```

The above listing shows two new mechanisms:

* `evt.data.source` – paste event's data can be extended to contain `source` property, which will indicate the source of paste data (Word in the example). It can be used to detect data format in other plugins.
* `editor._.nextPasteType` – this property will be used to force paste type. It can be added by paste commands (like `pastefromword` one) and removed at the end via generic handler in `pastetools`.

However making such handling generic forces us to refactor most of paste processing in our plugins, which is probably out of scope for 4.13.0.

Comandeer replied:
https://github.com/ckeditor/ckeditor-dev/issues/3369

Comandeer commented:
> Content in Safari like this one, after <kbd>cmd</kbd> + <kbd>a</kbd>, <kbd>cmd</kbd> + <kbd>c</kbd> and <kbd>cmd</kbd> + <kbd>v</kbd> in editor doesn't contain `guid-...` so might won't be recognized as content from google docs. It might be importnatn if there is no general content fixer regardless of data source.

We have very similar issue with Excel in Safari – https://github.com/ckeditor/ckeditor-dev/issues/2609#issuecomment-439917031
I've reported bug to Safari: https://bugs.webkit.org/show_bug.cgi?id=200569

Comandeer commented:
> Few test for build versions fails on CI (check recent Travis build)

As errors were nonsensical (claiming that basic pastetools methods aren't defined), I triggered a custom build → https://travis-ci.org/ckeditor/ckeditor-dev/builds/569799069

It seems that Travis is unstable again 😞

> some unit tests are failing on Edge

Can't reproduce.

Comandeer commented:
> When it comes to generic paste manual test I would add some nested lists to sample GDocs document.

I've updated tests, however I'm not able to generate sensible fixtures for Edge and IE11 due to unexpected bug in GDocs (seems to be introduced recently): content is copied incorrectly, when there's is an image at its beginning (in fact, in Edge only the image is copied).

> And two manual tests related to loading filters fails on IE (checked on IE11 and IE8):

It was connected with the fact that IE does not expose proper MIME type of pasted content. I relaxed the conditional to fire paste handler.

> One more thing, do we assume manual test with pasting from GDocs could be also run on browsers which doesn't support GDocs itself? If so it should be mentioned that you could open the document in different browser (however, I am not sure if specific browser would not somehow affect final result).

I've added the note. Probably more troublesome is copying in such browser, not pasting (as the pasted HTML will be the same as in modern browser).

I've also added handling legacy config names.

I haven't touch nested lists normalizations yet.

Comandeer commented:
I've fixed issue with pasting fragments of nested lists.

Comandeer commented on tests/plugins/pastefromgdocs/generated/basic.js:
```diff
@@ -0,0 +1,104 @@
+					'test Basic gdocs firefox': CKEDITOR.env.ie && CKEDITOR.env.version <= 11,
+					'test Basic gdocs safari': CKEDITOR.env.ie && CKEDITOR.env.version <= 11,
+					// 'test Basic gdocs edge': CKEDITOR.env.ie && CKEDITOR.env.version <= 11,
+					// 'test Basic gdocs ie11': !CKEDITOR.env.ie || CKEDITOR.env.version > 11
```
https://github.com/ckeditor/ckeditor-dev/pull/3257#issuecomment-520240791

> I've updated tests, however I'm not able to generate sensible fixtures for Edge and IE11 due to unexpected bug in GDocs (seems to be introduced recently): content is copied incorrectly, when there's is an image at its beginning (in fact, in Edge only the image is copied).

Comandeer replied:
It seems that GDocs resolved the issue, so I updated fixtures.

Comandeer commented:
> CI fails for build editor (one test checking handler priorities)

Built version added `pastefromword` filters to the mix, causing the fail of assertions. I've removed `pastefromword` plugin.

> Also, some unit tests still fails on Edge for me:

It seems that Edge has serious issues with cache 😐 After refreshing test page for several times, they started passing.

> In IE11 the first image is not centered after paste:

Unfortunately it's impossible to fix, as the info about image being centered is lost during copying.

> Copying nested lists fails on IE8:

This issue was super easy to fix. IE8 just strips quotes from around `[id]` attribute we use to detect GDocs content. Making quotes optional in our regex fixed the issue.

> Some JSDuck errors during docs building are also visible:

Fixed.

Comandeer commented on plugins/pastefromgdocs/filter/default.js:
```diff
@@ -0,0 +1,136 @@
+		rules: [
+			commonFilter.rules,
+			CKEDITOR.plugins.pastetools.filters.gdocs.rules
+		]
```
Filters are loaded once. To load filters you must pass their URLs into `filters` array inside `editor.pasteTools.register` and then they can be used also inside other filters.

f1ames commented:
@Comandeer could you take a look on the CI. One test fails for built version - https://travis-ci.org/ckeditor/ckeditor-dev/jobs/579456716#L625, looks like some minor thing in recently introduced unit test.

Comandeer commented:
I've fixed failing test and rebased branch onto latest `major`.

f1ames commented:
Rebased onto latest `major`.

Comandeer commented:
The bug with list is not fixable on our side… GDocs just pass such data to us: last list item is outside the list. Sample data: https://gist.github.com/Comandeer/d851b67e47c81baeef1e416dc4e1521b

As you can see, the last element is after `</ul>`, enclosed inside `span`.

There is also no way to detect such situation, as there are no signs that such `span` is from the list. Even if we assume that all `span`s after lists are in fact last list elements (which is a very silly assumption), we still don't know, on which level of nesting the last item was.

I've updated docs with deprecated info for both `CKEDITOR.plugins.pastefromword` namespace and several PfW config variables. I've also updated `@since` tags for elements in `CKEDITOR.plugins.pastetools` as it didn't make any sense to have something added in 4.13.0 with `@since 3.1.0` tag. However there is one issue with such approach:

<img width="775" alt="CKEDITOR.config.pasteTools_removeFontStyles added and deprecated in 4.13.0" src="https://user-images.githubusercontent.com/1078728/64880919-b3e3e080-d659-11e9-8d10-47e346dbf271.png">

Despite this little issue, I feel that the current shape of docs is much more readable and sensible than before.

LiamDawe commented:
So how do we get this to work correctly? I've upgraded to the latest release and when pasting in firefox, there's a lot of extra paragraph spacing where there shouldn't be.
